### ISSUE TITLE
Add PIDfile support

### ISSUE THREAD

**Issue by dtouzeau** [m13253/dns-over-https#29]
In order to be compatible with "monit", is it possible to add the possibility to add the pidfile support in the doh-server ?

**Comment by m13253** [comment]
It is a good idea!
I have been using systemd so I never considered about pidfile.

But do you know how to do that? Do I just have to write my PID to a file and remove it before exiting?

Or, can it be done by a bash script like this?
```bash
#!/bin/bash
set -e
doh-server "$@" &
echo "$$" >/var/run/doh-server.pid
wait
```

Would you please give me any suggestions?

**Comment by dtouzeau** [comment]
No, inside your code
a piece of example, you just have to write the file eg /var/run/doh-server.pid

import (
         "os","fm"
 )

 func main() {

         pid := os.Getpid()

         parentpid := os.Getppid()

         fmt.Printf("The parent process id of %v is %v\n", pid, parentpid)

         proc, err := os.FindProcess(pid) // or replace with other process number

         if err != nil {
                 panic(err)
         }

         fmt.Println(proc.Pid)

 }

**Comment by m13253** [comment]
I am not sure about 2 things:
- Why do I need to care about Parent PID?
- Why do I need to FindProcess, and then get the PID? (since I already have the PID)

Would you please explain it to me?

**Comment by dtouzeau** [comment]
No " parentpid := os.Getppid()" you did not need it
was just an example
In fact you just need "pid := os.Getpid()" and put pid in the file
I'm not familiar with go but for me the real code is more simple:

     import (
         "fmt"
         "log"
         "os"
     )

     func main() {
         file, err := os.Create("/var/run/doh-server.pid")
         if err != nil {
             log.Fatal("Cannot create pid file file", err)
         }
         pid := os.Getpid()
         defer file.Close()
         fmt.Fprintf(file, pid)
     }

**Comment by m13253** [comment]
OK. Thank you for that!
And, do I need to remove the file before the program exits? Or just leave it as is?

**Comment by dtouzeau** [comment]
In fact 
You have to remove it if a signal "KILL" is sent to your daemon.

When your program start: 
Check if /var/run/doh-server.pid exists.
If exists then read the pid in /var/run/doh-server.pid
get the integer, and check if you have a /proc/[PID]/cmdline file.
if /proc/[PID]/cmdline exists then an already instance of doh-server exists then die()
if /proc/[PID]/cmdline did not exists, remove /var/run/doh-server.pid and create a new one with the new PID file.
This to prevent a script that loop to run multiple time your program.

**Comment by m13253** [comment]
Understood! Thank you for your detailed explanation!

Please give me some 1 or 2 days so I can implement and test the feature.
I will tell you when I finish. Please help me to test the stability then.

By the way, I found it a little bit difficult to implement on other OSes, due to the following reasons:
- On Windows, we don't have a global tmpfs which is erased after reboot.
- On macOS, we don't have either a global tmpfs and a procfs. And due to the documentation, os.FindProcess can't check whether a process is still alive on Unix systems.

However PID file is not necessary for Windows and macOS:
- On Windows, one of my friend is working on integration with NT Services. Thus PID file will not be implemented on Windows.
- On macOS, since we already have launchd integration, PID file will not be implemented either.

Therefore, PID file will only be implemented on Linux for daemon management systems that are not able to use cgroups to track processes (e.g. sysvinit / upstart / monit / some Dockerized environments). Is it enough for your scenario? Do I need to expand the support to other operating systems such as BSD?

After the feature is added to the code, you will be able to try the new feature in master branch.
But a version number and tarball downloads will be released only after I collect 2-3 new features.

**Comment by Sherlock-Holo** [comment]
i think if add pid file support, user a flag to turn on this feature,
because if it runs under systemd the pid file support is unnecessary

David Touzeau <notifications@github.com> 于 2018年11月24日周六 16:21写道：

**Comment by m13253** [comment]
Fixed in commit 35ddf435059fb3f4a9a2903a5235622fe3d3a590.
Please have a test.

The feature will only be available on certain platforms.

**Comment by m13253** [comment]
Fixed, closing.
